### ISSUE TITLE
Investigate UOP Accessibility Issues

### ISSUE THREAD

**Issue by Scott-James-Hurley** [UserOfficeProject/user-office-project-issue-tracker#1589]
We should use the [Web Content Accessibility Guidelines](https://www.w3.org/WAI/standards-guidelines/wcag/) to identify issues in UOP. Useful tools for helping spot problems are [WAVE](https://wave.webaim.org/extension/), [ARIA DevTools](https://chromewebstore.google.com/detail/aria-devtools/dneemiigcbbgbdjlcdjjnianlikimpck?hl=en) and a screen reader like [NVDA](https://www.nvaccess.org/download/). This investigation could be carrier out on a by page or by use case basis. After discussion, I'll update with issue with a checklist of pages/use cases to check.

**Comment by Scott-James-Hurley** [comment]
related: https://github.com/UserOfficeProject/issue-tracker/issues/139

**Comment by Scott-James-Hurley** [comment]
I went through each page as a User Officer (all roles share components - it's only the data populating the components that changes) and noted all the issues flags by WAVE. The issues raised below should be evaluated and prioritised according to how much they affect usability and how difficult they would be to resolve.

Assigned Roles Modal:
- Skipped Heading Level: h2 -> h6, this is an inherent property of MaterialTable
- Broken ARIA reference: `user-office-core/apps/frontend/src/components/AppToolbar/AccountActionButton.tsx` line 92 aria-labelledby should point to element

Root as User Officer:
- Missing Label: `user-office-core/apps/frontend/src/components/common/proposalFilters/CallFilter.tsx`: element id 'call-select' input is missing a corresponding form label
- Material UI seems to have some inputs without labels and contrast issues with the same elements (questionable)
- Missing or uninformative page title
- Missing Table Header - make sure all tables have header

Calls as User Officer:
- Status label not linked to dropdown
- Rows per page label not linked to dropdown
- Empty table header

Technique Proposals as User Officer:
- "Welcome to the Xpress proposal management page" - not a heading
- Call filter input is missing a corresponding form label
- Technique filter input is missing a corresponding form label
- Experimental area filter input is missing a corresponding form label
- Status filter input is missing a corresponding form label
- Contrast and missing label issues for instrument column values in table - investigate
- Rows per page dropdown missing label
- The supplemental information for the Experimental Area and Status columns are only accessible by mouse
- The table has an empty column somewhere
- Select All Rows doesn't have a header
- There are contrast issues reported by WAVE but I don't see them?
- "These can be edited for status, full proposal information can be viewed in this system and here" - make the link text more descriptive

Faps as User Officer:
- Status filter label is orphaned
- Rows per page dropdown is missing a label
- Review visibility label orphaned from input

FapPage as User Officer:
General:
- aria-labelledby="horizontal-tabpanel-0" but element is horizontal-tab-0

Proposals and Assignments:
- Call and experimental area labels orphaned from inputs
- Checkbox for each table element missing a label
- The table has an empty header somewhere
- Select All Rows doesn't have a header
- Is the title attribute for proposal titles needed? Maybe for styling?
- Call filter label orphaned from input
- Rows per page missing label

Meeting Components:
- Call filter label orphaned from input
- Rows per page missing label

Legacy Proposals:
- Call and experimental area labels orphaned from inputs
- Checkbox for each table element missing a label
- The table has an empty header somewhere
- Select All Rows doesn't have a header
- Is the title attribute for proposal titles needed? Maybe for styling?

Instruments as User Officer:
- Rows per page dropdown missing label

Assign experiment scientist modal:
- simple-modal-title heading contains no text
- Close button has no text - not visible to screen readers?

Edit instrument modal:
- Is there an extraneous hidden field?

Techniques as User Officer:
- Empty table header
- Rows per page dropdown is missing a label
- Do the technique descriptions need titles?

Create technique modal:
- Is there an extraneous hidden field?

Assign scientist:
- Empty heading
- Close button has no text

Tag as User Officer:
- Instruments expanded - Instruments text skips heading level h2 -> h6
- Rows per page dropdown is missing a label

Add instrument:
- Empty heading
- Close button has no text
- Instrument checkboxes have no label
- Rows per page dropdown is missing a label
(Assign call has the same issues)

Features as User Officer:
- "Features" heading skips levels
- Table header contains no text

Settings as User Officer
- "App Settings" header skips levels

View Proposals as User Officer:
- Material UI empty heading
- "New Proposal" heading level skip h2 -> h6
- "finalstatus" input has no label

**Comment by Scott-James-Hurley** [comment]
Keyboard Navigation:
- Can't add users as co-i  when creating a proposal
- Can't seem to access all the text formatting options in text boxes with just keyboard
- Input validation errors should be more obvious 
- Maybe there should be an explicit edit button along with copy and delete for eg samples?
- Create/update call modal - cant access all fields with just keyboard
- Proposals table - can't add/remove columns
- Can't access FAP documents with keyboard
- Proposal workflows aren't usable with just keyboard but could be fixable: https://reactflow.dev/learn/advanced-use/accessibility
- Select File button from import template doesn't accept keystrokes
- Questionary editor - elements aren't highlighted, are we going to try to make this keyboard accessible?
- questionaryEditor - can't add new questions or edit existing ones
- Email template - need a way to exit text box for email body
- PDF template editor - name and description cancel and confirm buttons have no name so their functions are ambiguous, you can't edit name and description with just a keyboard and you need to be able to exit body text box
- Account menu can't be used with keyboard

I didn't try every proposal question but a decent variety

**Comment by Scott-James-Hurley** [comment]
The Accessibility Insights for Web tool flagged the following:
- Instrument and status filters for proposals table should have aria-label and aria-roles attributes
- Table column header elements need aria parent role: row
- FAP General tab - Review Visibility input field issue: https://accessibilityinsights.io/info-examples/web/aria-input-field-name/
- Table expand controls issue: https://accessibilityinsights.io/info-examples/web/label/
- Close person modal button issue: https://accessibilityinsights.io/info-examples/web/button-name/
- Close instrument/call modal button issue on tags page: https://accessibilityinsights.io/info-examples/web/button-name/, also doesn't have table title
- Proposal workflow editor: "Connection line style" text low contrast, list of available statuses - has children that are not allowed: [role=button]
- Email template editor textbox issue: https://accessibilityinsights.io/info-examples/web/aria-input-field-name/
- PDF template editor textbox issue: https://accessibilityinsights.io/info-examples/web/aria-input-field-name/
- Edit People title dropdown issue: https://accessibilityinsights.io/info-examples/web/aria-input-field-name/
- Edit Question read permissions dropdown issue: https://accessibilityinsights.io/info-examples/web/aria-input-field-name/
- Not all modals have an exit button
- Currently selected tab - Invalid ARIA attribute value: aria-controls="horizontal-tabpanel-0" (https://accessibilityinsights.io/info-examples/web/aria-valid-attr-value/) e.g.
<img width="245" height="159" alt="Image" src="https://github.com/user-attachments/assets/b65cec05-ac08-4f7d-aa3e-65a469c10eb9" />

**Comment by Scott-James-Hurley** [comment]
Issues I found when using a screen reader:

- For a lot of modals (not all), focus is not shifted to the modal and elements in the background aren't hidden
- Proposals page - the call filter combobox is the only landmark (don't know if it should be one) but the other filters aren't and the table isn't a landmark
- Proposal review - if there's no selectable elements in the tab e.g. there are no technical reviews, the text is not read out to a user using just a keyboard
- Date input field should read label (e.g. start and end date) when you select it so you know the context
- You can't exit the search box in assign instruments to selected call modal
- First row of assign instruments to selected call modal (Calls page) - screen reader repeats "FAP" a lot - not the case for other rows
- Text can stop being read at seemingly arbitrary points - perhaps there's a new line in the text that doesn't show up in the UI? example: reference number format tooltip, you need to select the next element (down arrow) to continue reading
- FAP page - General tab- Custom Grade Guide checkbox - textbox that appears after checking the box does not get focus, Code -> FAP Code, Description -> FAP Description, Active checkbox -> Is FAP active (fields should be a bit more self-descriptive)
- FAP - Members - add chair button selection focus should come after fap chair field focus and add secretary button focus should come after secretary field focus
- FAP - Documents - should be able to tab between document download links
- FAP - Meeting Component - you can't reorder proposals with just a keyboard
- Assign instrument to tag modal - The table title isn't descriptive and the checkbox comes before name of instrument - this could be confusing for screen reader users as they don't know what the checkbox relates to
- Assign call to tag modal - The table title isn't descriptive and the checkbox comes before name of instrument - this could be confusing for screen reader users as they don't know what the checkbox relates to
- Proposal workflow editor - status events and actions modal - next status events checkboxes label isn't read out when the field is focused
- Edit proposal workflow - workflow properties fields (title, description) should have labels
- Edit question modal - add dependency fields labels aren't not read out, same with read permissions
- Drag and drop is a problematic for keyboard users and screen readers
- Email status action logs - viewing a proposal then exiting the view should bring you back to the status action logs page
- Edit person page - no header which is inconsistent  with other pages, title combobox label not read out
- Create/update API key - the checkbox label isn't read out when it's highlighted, copy button should explain it's copying the key, make the update button easier to access? - add one at the top of the list of permissions as well? add a permissions count so screen reader users are less likely to get lost?
- Role management - The "Role Management" header is level 4 but the "Existing Roles" header is level 2 which is lower and smaller
- Should the Features page have an update button to be consistent?
- App settings - highlighting the edit button should also read out the setting you are editing

**Comment by Scott-James-Hurley** [comment]
Will open again if new issues are found.
